middle_end/flambda/terms/let_expr.rec.ml

```diff
@@ -30,10 +30,43 @@ let pattern_match t ~f =
   A.pattern_match t.name_abstraction
     ~f:(fun bindable_let_bound body -> f bindable_let_bound ~body)
 
-let pattern_match_pair t1 t2 ~f =
-  A.pattern_match_pair t1.name_abstraction t2.name_abstraction
-    ~f:(fun bindable_let_bound body1 body2 ->
-      f bindable_let_bound ~body1 ~body2)
+module Pattern_match_pair_error = struct
+  type t = Mismatched_let_bindings
+
+  let to_string = function
+    | Mismatched_let_bindings -> "Mismatched let bindings"
+end
+
+let pattern_match_pair t1 t2 ~dynamic ~static =
+  A.pattern_match t1.name_abstraction
+    ~f:(fun bindable_let_bound1 body1 ->
+      A.pattern_match t2.name_abstraction
+        ~f:(fun bindable_let_bound2 body2 ->
+          let dynamic_case () =
+            let ans =
+              A.pattern_match_pair
+                t1.name_abstraction
+                t2.name_abstraction
+                ~f:(fun bindable_let_bound body1 body2 ->
+                  dynamic bindable_let_bound ~body1 ~body2)
+            in
+            Ok ans
+          in
+          match bindable_let_bound1, bindable_let_bound2 with
+          | Bindable_let_bound.Singleton _,
+            Bindable_let_bound.Singleton _ ->
+            dynamic_case ()
+          | Set_of_closures { closure_vars = vars1; _ },
+            Set_of_closures { closure_vars = vars2; _ } ->
+            if List.compare_lengths vars1 vars2 = 0
+            then dynamic_case ()
+            else Error Pattern_match_pair_error.Mismatched_let_bindings
+          | Symbols bound_symbols1,
+            Symbols bound_symbols2 ->
+            let ans = static ~bound_symbols1 ~bound_symbols2 ~body1 ~body2 in
```

Comment:
Is it expected that the callback is called if there are different numbers of bound symbols in each of the lets?

Reply:
Good point. Adding a check.